### DISCUSSION TITLE
api.Println working differently sometimes

### DISCUSSION THREAD

**Discussion by infinite-pursuits** [Consensys/gnark#923] [Q&A]
when I use api.Println, sometimes negative numbers are printed as "-102" for instance but other times it is printed as a big number like  "21888242871839275222246405745257275088548364400416034343698204186575807914052" which i believe is also a negative number. Why are negative numbers printed differently?

**Comment by gbotrel** [comment]
Hi -- there is no concept of "negative number" in finite fields; all modern SNARK constructions work with finite fields ("big integer modulo a prime").

**Reply by infinite-pursuits** [reply]
yes but why do I see something like  "-102" being printed?

**Reply by ivokub** [reply]
Test engine handles `Println` differently than the actual solver. If variable, represented as a negative value (i.e. value-modulus) is small, then we print it as a negative value. I agree, it makes debugging more difficult.

You're welcome to modify the implementation and create a PR. For me only interesting negative case would be "-1", then some obvious cases are more apparent. We could discuss further in PR comments, I see that @Tabaie has touched that function lastly.

See the implementation https://github.com/Consensys/gnark/blob/master/test/engine.go#L514-L522
